pylabrobot/liquid_handling/liquid_handler.py

```diff
@@ -738,8 +752,24 @@ async def aspirate(
     for op in aspirations:
       if does_volume_tracking():
         if not op.resource.tracker.is_disabled:
+
+          # Cross contamination check
+          if does_cross_contamination_tracking():
+            if check_clean(next(reversed(op.liquids))[0], op.tip.tracker.liquid_history):
+              raise CrossContaminationError(
+                f"Attempting to aspirate {next(reversed(op.liquids))[0]} with \
+                    a tip contaminated with {op.tip.tracker.liquid_history}.")
+
           op.resource.tracker.remove_liquid(op.volume)
         for liquid, volume in reversed(op.liquids):
+
+          # Cross contamination check
+          if does_cross_contamination_tracking():
+            if check_clean(liquid, op.tip.tracker.liquid_history):
+              raise CrossContaminationError(
+                f"Attempting to aspirate {next(reversed(op.liquids))[0]} with \
+                    a tip contaminated with {op.tip.tracker.liquid_history}.")
+
```

Comment:
Could you explain what's going on here?

You loop over the liquids in the aspiration, and call check_clean for each. It seems that in the loop it is checked that none of the liquids in the aspiration have been in the tip before. (If the tip has seen any liquid at all.) If you aspirated some liquid with this tip before, and then aspirate from the same container again, does this (incorrectly?) raise an error?

There seems to be duplication with line 758, where this is checked only for the last liquid in the aspiration.

I thought cross contamination checking would check that the tip has not touched liquids that are not in the well/container. This would mean looping over the liquids in the tip and ensuring there are none which are not also in the well. If the liquid history of the well and tip are represented as sets, this can be done by checking that the `tip_history` is a strict subset of `well_history`.

Reply:
> You loop over the liquids in the aspiration, and call check_clean for each. It seems that in the loop it is checked that none of the liquids in the aspiration have been in the tip before. (If the tip has seen any liquid at all.) If you aspirated some liquid with this tip before, and then aspirate from the same container again, does this (incorrectly?) raise an error?

Your interpretation is correct. Check clean takes in a tip and returns true if the candidate aspiration liquid is not in the liquid history AND if liquid history is not empty. We want check_clean to return false for a successful aspiration. I think a better name for this would be check_contaminated. We only want the check to fire if the candidate liquid is not already in the liquid history, but only if the liquid_history is not empty. Imagine a tip touches red dye. Red dye is now stored in the liquid_history. If we want to aspirate from a well containing blue dye, we check that blue dye is not in liquid_history. If it is, and the liquid history isn't empty, then the tip is allowed to aspirate, so the check_clean returns false. Otherwise, we return true and except. This does not return an error if we aspirate from the same container twice.

> There seems to be duplication with line 758, where this is checked only for the last liquid in the aspiration.

I think you only need one check in either spot. I removed the first one. We do the check for every liquid entering the tip.

> I thought cross contamination checking would check that the tip has not touched liquids that are not in the well/container. This would mean looping over the liquids in the tip and ensuring there are none which are not also in the well. If the liquid history of the well and tip are represented as sets, this can be done by checking that the `tip_history` is a strict subset of `well_history`.

Since we do the check for every liquid entering the tip, I believe that is the same as checking the strict subset condition.